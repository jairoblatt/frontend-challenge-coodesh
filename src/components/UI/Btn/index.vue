<template>
  <button v-on="$listeners" :class="classes">
    <spinner v-show="loading" size="21" />
    <slot v-show="!loading" />
  </button>
</template>

<script lang="ts">
import Vue, { PropType } from "vue";

export default Vue.extend({
  name: "Btn",

  props: {
    primary: Boolean as PropType<boolean>,
    secondary: Boolean as PropType<boolean>,
    block: Boolean as PropType<boolean>,
    loading: Boolean as PropType<boolean>,
    small: Boolean as PropType<boolean>,
    outlined: Boolean as PropType<boolean>,
  },

  computed: {
    classes() {
      return {
        btn: true,
        "btn--primary": this.primary,
        "btn--secondary": this.secondary,
        "btn--block": this.block,
        "btn--small": this.small,
        "btn--outlined": this.outlined,
      };
    },
  },
});
</script>

<style lang="postcss" scoped>
.btn {
  @apply rounded-md px-5 py-3 text-sm outline-none font-semibold text-white transition-colors;
}

.btn--primary {
  @apply bg-primary-base hover:bg-primary-light;
}

.btn--secondary {
  @apply bg-secondary-gray;
}

.btn--block {
  @apply w-full;
}

.btn:disabled {
  @apply bg-gray-300 text-gray-900 cursor-auto;
}

.btn--small {
  @apply px-7 py-2;
}

.btn--outlined {
  @apply border border-gray-100 text-light-onSurfacePrimary hover:border-black hover:text-black;
}
</style>
