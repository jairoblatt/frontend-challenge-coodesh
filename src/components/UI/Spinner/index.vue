<template>
  <div class="spinner-container">
    <span class="spinner" :style="styles"></span>
  </div>
</template>
<script lang="ts">
import Vue, { PropOptions } from "vue";
import CSS from "csstype";

export default Vue.extend({
  name: "Spinner",

  props: {
    size: {
      type: [String, Number],
      default: "inherit",
    } as PropOptions<string | number>,
  },

  computed: {
    styles(): CSS.Properties {
      return {
        minHeight: `${this.size}px`,
        minWidth: `${this.size}px`,
      };
    },
  },
});
</script>
<style lang="postcss" scoped>
.spinner-container {
  @apply relative flex justify-center items-center min-h-full min-w-full;
}

.spinner {
  @apply pointer-events-none border-solid rounded-full  border-primary-base  p-1;

  border-width: 0.3rem;
  border-top-color: white;
  animation: spinner 0.5s linear infinite;
}

@keyframes spinner {
  100% {
    transform: rotate(360deg);
  }
}
</style>
