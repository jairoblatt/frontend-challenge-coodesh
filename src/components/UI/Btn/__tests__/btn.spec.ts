import { mount, MountOptions, Wrapper, createLocalVue } from "@vue/test-utils";
import Btn from "../index.vue";
import Spinner from "@/components/UI/Spinner/index.vue";

describe("Btn.vue", () => {
  type Instance = InstanceType<typeof Btn>;
  let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>;

  const localVue = createLocalVue();
  localVue.component("Spinner", Spinner);

  beforeEach(() => {
    mountFunction = (config = {}) => {
      return mount(Btn, {
        localVue,
        ...config,
      });
    };
  });

  it("Should mount the componente", () => {
    const wrapper = mountFunction();

    expect(wrapper.vm).toBeDefined();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Should display the content passed by slot", () => {
    const wrapper = mountFunction({
      slots: {
        default: "My btn here",
      },
    });

    expect(wrapper.text()).toBe("My btn here");
  });

  it('Should have the "btn--primary" class when the prop is passed', () => {
    const wrapper = mountFunction({
      propsData: {
        primary: true,
      },
    });

    expect(wrapper.classes("btn--primary")).toBe(true);
  });

  it('Should have the "btn--secondary" class when the prop is passed', () => {
    const wrapper = mountFunction({
      propsData: {
        secondary: true,
      },
    });

    expect(wrapper.classes("btn--secondary")).toBe(true);
  });

  it('Should have the "btn--block" class when the prop is passed', () => {
    const wrapper = mountFunction({
      propsData: {
        block: true,
      },
    });

    expect(wrapper.classes("btn--block")).toBe(true);
  });

  it('Should emit "click" event when is clicked', () => {
    const onClick = jest.fn();
    const wrapper = mountFunction({
      listeners: {
        click: onClick,
      },
    });

    wrapper.find("button").trigger("click");

    expect(onClick).toHaveBeenCalled();
  });
});
