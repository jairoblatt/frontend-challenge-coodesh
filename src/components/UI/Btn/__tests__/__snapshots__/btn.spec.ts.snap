// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Btn.vue Should mount the componente 1`] = `
"<button class=\\"btn\\">
  <div class=\\"spinner-container\\" style=\\"display: none;\\"><span class=\\"spinner\\" style=\\"min-height: 21px; min-width: 21px;\\"></span></div>
</button>"
`;
